File: frontend/mobile/lib/core/services/preferences_service.dart
import 'package:shared_preferences/shared_preferences.dart';
import 'dart:convert';

/// Service for managing user preferences
class PreferencesService {
  static const String _keyOnboardingCompleted = 'onboarding_completed';
  static const String _keyMadhab = 'madhab';
  static const String _keyTheme = 'theme';
  static const String _keyFontSize = 'font_size';
  static const String _keyEnableAnimations = 'enable_animations';
  static const String _keyEnableNotifications = 'enable_notifications';
  static const String _keyLanguage = 'language';
  static const String _keyAllPreferences = 'all_preferences';

  final SharedPreferences _prefs;

  PreferencesService(this._prefs);

  // Onboarding
  Future<void> setOnboardingCompleted(bool completed) async {
    await _prefs.setBool(_keyOnboardingCompleted, completed);
  }

  bool getOnboardingCompleted() {
    return _prefs.getBool(_keyOnboardingCompleted) ?? false;
  }

  // Madhab
  Future<void> setMadhab(String madhab) async {
    await _prefs.setString(_keyMadhab, madhab);
  }

  String? getMadhab() {
    return _prefs.getString(_keyMadhab);
  }

  // Theme
  Future<void> setTheme(String theme) async {
    await _prefs.setString(_keyTheme, theme);
  }

  String getTheme() {
    return _prefs.getString(_keyTheme) ?? 'light';
  }

  // Font Size
  Future<void> setFontSize(String fontSize) async {
    await _prefs.setString(_keyFontSize, fontSize);
  }

  String getFontSize() {
    return _prefs.getString(_keyFontSize) ?? 'medium';
  }

  // Animations
  Future<void> setEnableAnimations(bool enable) async {
    await _prefs.setBool(_keyEnableAnimations, enable);
  }

  bool getEnableAnimations() {
    return _prefs.getBool(_keyEnableAnimations) ?? true;
  }

  // Notifications
  Future<void> setEnableNotifications(bool enable) async {
    await _prefs.setBool(_keyEnableNotifications, enable);
  }

  bool getEnableNotifications() {
    return _prefs.getBool(_keyEnableNotifications) ?? true;
  }

  // Language
  Future<void> setLanguage(String language) async {
    await _prefs.setString(_keyLanguage, language);
  }

  String getLanguage() {
    return _prefs.getString(_keyLanguage) ?? 'ar';
  }

  // Get all preferences as JSON
  Map<String, dynamic> getAllPreferences() {
    return {
      'onboarding_completed': getOnboardingCompleted(),
      'madhab': getMadhab(),
      'theme': getTheme(),
      'font_size': getFontSize(),
      'enable_animations': getEnableAnimations(),
      'enable_notifications': getEnableNotifications(),
      'language': getLanguage(),
    };
  }

  // Save all preferences from JSON
  Future<void> setAllPreferences(Map<String, dynamic> preferences) async {
    if (preferences.containsKey('onboarding_completed')) {
      await setOnboardingCompleted(preferences['onboarding_completed']);
    }
    if (preferences.containsKey('madhab')) {
      await setMadhab(preferences['madhab']);
    }
    if (preferences.containsKey('theme')) {
      await setTheme(preferences['theme']);
    }
    if (preferences.containsKey('font_size')) {
      await setFontSize(preferences['font_size']);
    }
    if (preferences.containsKey('enable_animations')) {
      await setEnableAnimations(preferences['enable_animations']);
    }
    if (preferences.containsKey('enable_notifications')) {
      await setEnableNotifications(preferences['enable_notifications']);
    }
    if (preferences.containsKey('language')) {
      await setLanguage(preferences['language']);
    }
  }

  // Backup preferences to JSON string
  String backupPreferences() {
    final prefs = getAllPreferences();
    return jsonEncode(prefs);
  }

  // Restore preferences from JSON string
  Future<void> restorePreferences(String jsonString) async {
    try {
      final Map<String, dynamic> prefs = jsonDecode(jsonString);
      await setAllPreferences(prefs);
    } catch (e) {
      throw Exception('Failed to restore preferences: $e');
    }
  }

  // Reset all preferences to defaults
  Future<void> resetToDefaults() async {
    await _prefs.clear();
  }

  // Sync preferences with backend (placeholder for future implementation)
  Future<void> syncWithBackend() async {
    // TODO: Implement backend sync
    // This will be implemented when backend integration is ready
    final prefs = getAllPreferences();
    // await apiService.syncPreferences(prefs);
  }
}

File: frontend/nextjs-app/src/lib/services/preferences-service.ts
/**
 * Service for managing user preferences in localStorage
 */

export interface UserPreferences {
  onboardingCompleted: boolean;
  madhab?: string;
  theme: 'light' | 'dark' | 'auto';
  fontSize: 'small' | 'medium' | 'large' | 'xlarge';
  enableAnimations: boolean;
  enableNotifications: boolean;
  language: 'ar' | 'en';
}

const DEFAULT_PREFERENCES: UserPreferences = {
  onboardingCompleted: false,
  theme: 'light',
  fontSize: 'medium',
  enableAnimations: true,
  enableNotifications: true,
  language: 'ar',
};

const STORAGE_KEY = 'userPreferences';

export class PreferencesService {
  /**
   * Get all user preferences
   */
  static getPreferences(): UserPreferences {
    if (typeof window === 'undefined') {
      return DEFAULT_PREFERENCES;
    }

    try {
      const stored = localStorage.getItem(STORAGE_KEY);
      if (!stored) {
        return DEFAULT_PREFERENCES;
      }

      const parsed = JSON.parse(stored);
      return { ...DEFAULT_PREFERENCES, ...parsed };
    } catch (error) {
      console.error('Failed to load preferences:', error);
      return DEFAULT_PREFERENCES;
    }
  }

  /**
   * Save all user preferences
   */
  static setPreferences(preferences: Partial<UserPreferences>): void {
    if (typeof window === 'undefined') {
      return;
    }

    try {
      const current = this.getPreferences();
      const updated = { ...current, ...preferences };
      localStorage.setItem(STORAGE_KEY, JSON.stringify(updated));
    } catch (error) {
      console.error('Failed to save preferences:', error);
    }
  }

  /**
   * Get onboarding completion status
   */
  static getOnboardingCompleted(): boolean {
    return this.getPreferences().onboardingCompleted;
  }

  /**
   * Set onboarding completion status
   */
  static setOnboardingCompleted(completed: boolean): void {
    this.setPreferences({ onboardingCompleted: completed });
  }

  /**
   * Get madhab selection
   */
  static getMadhab(): string | undefined {
    return this.getPreferences().madhab;
  }

  /**
   * Set madhab selection
   */
  static setMadhab(madhab: string): void {
    this.setPreferences({ madhab });
  }

  /**
   * Get theme preference
   */
  static getTheme(): 'light' | 'dark' | 'auto' {
    return this.getPreferences().theme;
  }

  /**
   * Set theme preference
   */
  static setTheme(theme: 'light' | 'dark' | 'auto'): void {
    this.setPreferences({ theme });
  }

  /**
   * Get font size preference
   */
  static getFontSize(): 'small' | 'medium' | 'large' | 'xlarge' {
    return this.getPreferences().fontSize;
  }

  /**
   * Set font size preference
   */
  static setFontSize(fontSize: 'small' | 'medium' | 'large' | 'xlarge'): void {
    this.setPreferences({ fontSize });
  }

  /**
   * Get animations preference
   */
  static getEnableAnimations(): boolean {
    return this.getPreferences().enableAnimations;
  }

  /**
   * Set animations preference
   */
  static setEnableAnimations(enable: boolean): void {
    this.setPreferences({ enableAnimations: enable });
  }

  /**
   * Get notifications preference
   */
  static getEnableNotifications(): boolean {
    return this.getPreferences().enableNotifications;
  }

  /**
   * Set notifications preference
   */
  static setEnableNotifications(enable: boolean): void {
    this.setPreferences({ enableNotifications: enable });
  }

  /**
   * Get language preference
   */
  static getLanguage(): 'ar' | 'en' {
    return this.getPreferences().language;
  }

  /**
   * Set language preference
   */
  static setLanguage(language: 'ar' | 'en'): void {
    this.setPreferences({ language });
  }

  /**
   * Export preferences as JSON string for backup
   */
  static exportPreferences(): string {
    const prefs = this.getPreferences();
    return JSON.stringify(prefs, null, 2);
  }

  /**
   * Import preferences from JSON string
   */
  static importPreferences(jsonString: string): boolean {
    try {
      const prefs = JSON.parse(jsonString);
      this.setPreferences(prefs);
      return true;
    } catch (error) {
      console.error('Failed to import preferences:', error);
      return false;
    }
  }

  /**
   * Reset all preferences to defaults
   */
  static resetToDefaults(): void {
    if (typeof window === 'undefined') {
      return;
    }

    localStorage.removeItem(STORAGE_KEY);
  }

  /**
   * Sync preferences with backend (placeholder for future implementation)
   */
  static async syncWithBackend(): Promise<void> {
    // TODO: Implement backend sync
    // const prefs = this.getPreferences();
    // await apiClient.post('/api/preferences/sync', prefs);
  }
}
